extends Node

var should_change : bool = false

# Called when the node enters the scene tree for the first time.
func _ready():
	await get_tree().current_scene.ready
#	%ScrollContainer.get_tab_bar().focus_mode = Control.FocusMode.FOCUS_NONE
	held_sprite_is_null()
	Global.connect("reinfo", reinfo)
	Global.deselect.connect(held_sprite_is_null)
	


#region Update Slider info
func held_sprite_is_null():
	%SpriteID.text = "Sprite ID : 0"
	%Name.editable = false
	%Name.text = ""
	%NonAnimatedSheetCheck.disabled = true
	%FrameSpinbox.editable = false
	%AdvancedLipSync.disabled = true

func held_sprite_is_true():
	Global.top_ui.get_node("%DeselectButton").show()
	%Name.editable = true
	for i in Global.held_sprites:
		if i != null && is_instance_valid(i):
			if i.sprite_type == "Sprite2D":
				%AdvancedLipSync.disabled = false
				%NonAnimatedSheetCheck.disabled = false
				%FrameSpinbox.editable = true
				%SpriteID.text = "Sprite ID : " + str(i.sprite_id)

func reinfo():
	held_sprite_is_null()
	should_change = false
	for i in Global.held_sprites:
		if i != null && is_instance_valid(i):
			%Name.text = i.sprite_name
			if i.sprite_type == "Sprite2D":
				%AdvancedLipSync.button_pressed = i.get_value("advanced_lipsync")
				%NonAnimatedSheetCheck.button_pressed = i.get_value("non_animated_sheet")
				%FrameSpinbox.value = i.get_value("frame")
				%FrameSpinbox.max_value = (i.get_node("%Sprite2D").hframes * i.get_node("%Sprite2D").vframes) - 1
	await get_tree().create_timer(0.01).timeout
	held_sprite_is_true()
	should_change = true

func _on_name_text_submitted(new_text):
	if Global.held_sprites.size() <= 1:
		Global.held_sprites[0].treeitem.set_text(0, new_text)
		Global.held_sprites[0].sprite_name = new_text
		Global.held_sprites[0].save_state(Global.current_state)
	else:
		for i in Global.held_sprites.size():
			Global.held_sprites[i].treeitem.set_text(0, new_text + str(i+1))
			Global.held_sprites[i].sprite_name = new_text + str(i+1)
			Global.held_sprites[i].save_state(Global.current_state)

	Global.spinbox_held = false
	%Name.release_focus()
#endregion

#region Advanced-LipSync
func _on_advanced_lip_sync_toggled(toggled_on):
	if should_change:
		for i in Global.held_sprites:
			if i != null && is_instance_valid(i):
				if i.sprite_type == "Sprite2D":
					i.sprite_data.advanced_lipsync = toggled_on
					i.sprite_data.animation_speed = 1
					if toggled_on:
						i.get_node("%Sprite2D").hframes = 6
					else:
						i.get_node("%Sprite2D").hframes = 1
					i.advanced_lipsyc()
					i.get_node("%Sprite2D/Grab").anchors_preset = Control.LayoutPreset.PRESET_FULL_RECT
					i.save_state(Global.current_state)
					Global.reinfo.emit()

func _on_advanced_lip_sync_mouse_entered():
	%AdvancedLipSyncLabel.show()

func _on_advanced_lip_sync_mouse_exited():
	%AdvancedLipSyncLabel.hide()
#endregion

func _on_name_focus_entered() -> void:
	Global.spinbox_held = true

func _on_name_focus_exited() -> void:
	Global.spinbox_held = false

func _on_non_animated_sheet_check_toggled(toggled_on: bool) -> void:
	if should_change:
		for i in Global.held_sprites:
			if i != null && is_instance_valid(i):
				if i.sprite_type == "Sprite2D":
					%FrameSpinbox.max_value = (i.get_node("%Sprite2D").hframes * i.get_node("%Sprite2D").vframes) - 1
					i.sprite_data.non_animated_sheet = toggled_on
					i.animation()
					if toggled_on:
						%FrameHBox.show()
					else:
						%FrameHBox.hide()
				else:
					%FrameHBox.hide()

func _on_frame_spinbox_value_changed(value: float) -> void:
	if should_change:
		for i in Global.held_sprites:
			if i != null && is_instance_valid(i):
				if i.sprite_type == "Sprite2D":
					%FrameSpinbox.max_value = (i.get_node("%Sprite2D").hframes * i.get_node("%Sprite2D").vframes) - 1
					i.sprite_data.frame = clamp(value, 0, (i.get_node("%Sprite2D").hframes * i.get_node("%Sprite2D").vframes) - 1)
					i.get_node("%Sprite2D").frame = clamp(value, 0, (i.get_node("%Sprite2D").hframes * i.get_node("%Sprite2D").vframes) - 1)


func _on_frame_spinbox_mouse_entered() -> void:
	if should_change:
		for i in Global.held_sprites:
			if i != null && is_instance_valid(i):
				if i.sprite_type == "Sprite2D":
					%FrameSpinbox.max_value = (i.get_node("%Sprite2D").hframes * i.get_node("%Sprite2D").vframes) - 1
